tests/main/layout-change/task.yaml

```diff
@@ -0,0 +1,51 @@
+summary: Ensure that replacing a bind with a symlink doesn't break an update
+
+details: |
+    Recently, the steam snap replaced several bindings with symlinks in the
+    layout part, and that resulted in users not being able to update it from
+    the "binded" version to the "symlinked" one because snapd complained that
+    it couldn't update the mount namespace. The solution consists on discard
+    the namespace and retry again.
+
+# effectively we can only build the snap using LXD + snapcraft on ubuntu systems
+# due to various bugs with other distros and snapcraft and/or LXD
+systems:
+    # snapcraft is a classic snap, can't run on ubuntu-core
+    - -ubuntu-core*
+    # LXD doesn't work properly on amazon linux, fedora 31, opensuse, or debian
+    # 9 to build the snap, it complains about fuse.snapfuse getting EPERM when
+    # snaps are installed inside the LXD container that snapcraft creates
+    - -amazon-linux*
+    - -debian-*
+    - -fedora-*
+    - -opensuse-*
+    # additionally on fedora 32 and 33, and arch linux which are all cgroupsv2
+    # now, LXD does not work either, with the service failing to start up at all
+    # when we install LXD
+    - -arch-*
+    # snapcraft does not support running on centos 7 due to errors with the
+    # locale, see https://bugs.launchpad.net/snapcraft/+bug/1922140
+    # a separate error occurs on centos 8, see https://bugs.launchpad.net/snapcraft/+bug/1922981
+    - -centos-*
+    # ubuntu-18.04-32: i386 is not supported by lxd
+    # TODO: enable i386 by using lxd 3.0, currently snapcraft is failing to get the
+    #       lxd image when it is trying to build snapd
+    - -ubuntu-18.04-32
+
+
+execute: |
+    echo Testing layout change
+    # build the test snapcrafts; one with bind, and the next with symlink
+    snap install snapcraft --classic
+    snap install lxd
+    lxd init --auto
+    cp snapcraft_old.yaml snapcraft.yaml
+    snapcraft
+    cp snapcraft_new.yaml snapcraft.yaml
+    snapcraft
+
+    snap install --dangerous layouttest_0.1_*.snap
+    # the snap must be run; not doing it won't trigger the bug
+    snap run layouttest
+
+    snap install --dangerous layouttest_0.2_*.snap
```

Comment:
I'm confused what is happening here, because if I try that snap directly and try to run its command I get:

$ snap install --dangerous layouttest_0.2_amd64.snap 
layouttest 0.2 installed
$ layouttest 
cannot update snap namespace: cannot create symlink in "/usr/share": existing file in the way
snap-update-ns failed with code 1

Reply:
@pedronis  Ok, I prepared two new test snaps, removing pieces from the steam snaps until they had the bare minimum to trigger the bug. Now each snap can be installed separately, but the bug is triggered if version 0.1 is already installed when upgrading to version 0.2.